[CI-5155] [CI-5131] - revise set up test intelligence

aimurphy opened the pull request:
Clarified the calculation of **Duration** in the report.
General revisions throughout based on UI changes, additional languages supported, etc.

# Harness Developer Pull Request
Thanks for helping us make the Developer Hub better. The PR will be looked at
by the maintainers. 

## What Type of PR is This?

- [ x] Issue
- [ ] Feature
- [x ] Maintenance/Chore

If tied to an Issue, list the Issue(s) here:

* *Issue(s)*

## House Keeping
Some items to keep track of. Screen shots of changes are optional but would help the maintainers review quicker. 

- [ x] Tested Locally
- [ ] *Optional* Screen Shoot.

aimurphy commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+The test report is comprised of the following sections:
+
 #### Test Execution Overview
 
-Displays the overview of Total Test, Tests Selected, Duration of the Selected Tests, and Time Saved.
+Displays an overview of **Total Tests**, number of **Selected Tests**, total **Duration** of all tests, and **Time Saved**.
+
+**Duration** reflects the sum of all test durations collected as-is from the JUnit report. These values corresponds with the CPU time, and it is possible that the sum of these times may exceed the total pipeline execution time.
```
@hemanthmantri This is the specific change for CI-5155/CI-5131. The rest of the changes are editorial or based on UI changes. Although you may also want to review the YAML at the end.

hemanthmantri replied:
The test duration indicates the CPU time taken for all the tests to complete. 
Should we mention that the pipeline execution time is wall clock time?

aimurphy replied:
Revised as: **Duration** reflects the sum of CPU time taken for all tests to complete. The values are collected as-is from the JUnit report, and they don't correspond with wall clock time. In contrast, the pipeline execution time is a measure of wall clock time. Therefore, it is possible that the **Duration** may exceed the pipeline execution time.

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+   7. **Build Arguments**: Input the [arguments](../../ci-technical-reference/configure-run-tests-step-settings.md) for the build tool. These are used as input for the chosen build tool.
+   8. **Test Report Paths**: Input one or more test report paths.
+4. Expand **Additional Configuration** and complete these fields:
+   1. **Packages**: Leave blank or provide a comma-separated list of source code package prefixes, such as `com.company., io.company.migrations`. If blank, Harness auto-detects the package.
```
packages.

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+4. Expand **Additional Configuration** and complete these fields:
+   1. **Packages**: Leave blank or provide a comma-separated list of source code package prefixes, such as `com.company., io.company.migrations`. If blank, Harness auto-detects the package.
+   2. **Test Annotations**: Provide a comma-separated list of test annotations used in unit testing. Any method with a specified annotation is treated as a test method. The defaults are: `org.junit.Test, org.junit.jupiter.api.Test, org.testing.annotations.Test`
+   3. **Namespaces**: For .NET C# only, supply a comma-separated list of namespaces prefixes that you want to test.
```
namespace prefixes

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+   1. **Name**: Input a unique name for the step.
+   2. **Description**: Input a description for the step.
+   3. **Container Registry**: Select a container registry connector. Harness pulls the image to run build commands on from this container registry, for example, DockerHub.
+   4. **Image**: The name of the Docker image to use for running the build commands. Enter a Fully Qualified Image Name (FQIN) when using a private container registry.
```
Can we somehow mention that steps 3 and 4 are optional when "Harness Cloud" infrastructure is used?

aimurphy replied:
Yes, I'll add that.

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+   :::caution
 
-Open the PR.
+   Make sure the PR will trigger all tests, such as pom.xml, non-Java, and so on.
```
Ensure the build triggered due to this PR runs all the tests.

aimurphy replied:
Changed to: Make sure that the build triggered by this PR runs all tests, such as pom.xml, non-Java, and so on.

hemanthmantri replied:
Remove "such as pom.xml, non-java and son on.". Those are not tests. 

"Make sure that the build triggered by this PR runs all the tests."

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
 
-Once the Pipeline is executed, click **Tests**. **Tests** shows the test report of the test you configured and ran. Harness publishes test results in JUnit format. 
+1. Wait while the pipeline executes. To monitor the build's progress, go to **Builds** and select the build that was started by the PR.
+2. After the build succeeds, on the build details page, select **Tests**. **Tests** shows the test report for the test you configured for the **Run Tests** step. Harness publishes test results in JUnit format.
```
Harness parses test reports only if they are in JUnit format. Please make sure the XML reports being generated in your tests are in JUnit format. Otherwise, the **Tests** tab will not show any tests.

aimurphy replied:
Changed to: In order for the **Tests** tab to show tests, your test reports must be in JUnit XML format. Harness parses test reports that are in JUnit XML format only.

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
 #### Test Selection Breakdown
 
-Displays the selection of Tests on the following metrics: changed code, changed tests, and new tests.
+Test Intelligence makes decisions about which tests to run by assessing changes to the pipeline's codebase or the individual tests.
```
... by assessing changes to the code base (source files or test files).

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
 
 Displays the detailed list of all tests: class methods and test methods.
 
+You can filter the list to only show failed tests; sort the list by failure rate, duration, and total tests; and expand test suites to see details about individual tests in that suite.
```
The default view only shows failed tests. Users can toggle the "Show all Tests" switch to see all the tests.

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+                          connectorRef: Kubernetes_Quickstart
+                          namespace: harness-delegate
+              variables: []
+    projectIdentifier: CD_Examples
```
CI_Examples

aimurphy commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,185 @@ helpdocs_is_private: false
+Test Intelligence (TI) improves test time by running only the tests required to confirm the quality of the code changes that triggered the build. To learn more about how Test Intelligence works,go to [Test Intelligence Overview](../../ci-quickstarts/test-intelligence-concepts.md).
 
-In this topic, we'll cover how to set up Test Intelligence in a Harness CI Stage.
+You can enable Test Intelligence for builds in CI pipeline stages that use the **Run Tests** step. This topic explains how to set up Test Intelligence in a Harness CI pipeline stage.
```
Changed to "You can enable Test Intelligence by including a **Run Tests** step in your CI pipeline **Build** stages.

hemanthmantri commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+   :::caution
 
-Open the PR.
+   Make sure that the build triggered by this PR runs all tests, such as pom.xml, non-Java, and so on.
```
Remove    ", such as pom.xml, non-Java, and so on."

aimurphy replied:
removed

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+- **Updated Tests**: The number of tests that Test Intelligence ran because the tests themselves were changed
 
 #### Test Execution
 
 Displays the detailed list of all tests: class methods and test methods.
 
+Initially, the list shows only failed tests. To see all tests, toggle **Show all Tests**.
+
+You can sort the list by failure rate, duration, and total tests. You can also expand test suites to see details about individual tests in that suite.
+
 #### Call Graph
 
-A call graph is generated once the bootstrap is completed. In each later run, the test tab will display stats corresponding to the selection. 
+A call graph is generated after completing the bootstrap.
 
-#### Visualization
```
Use tabs instead of headings to present the various parts of the sections of the test report/

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
-### Optional: YAML Configure As Code
-
-Here's a working example of Test Intelligence on the open-source project Dubbo. Use this YAML template to create and configure the TI pipeline on the go.
+### Reference: Test Intelligence pipeline configuration as code
```
Remove "Reference:". Think you should add "example" to the end of the title since this shows an example yaml template.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
````diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
     tags: {}
 ```
-### See Also
+### See also
````
Remove this section. There's already a link to the overview at the beginning of the topic and there's no reason to link to view test reports here.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+### Perform the bootstrap
 
-Bootstrapping helps generate the initial call graph for Test Intelligence. Once the call graph is generated, TI can perform Test Selection. Perform the bootstrap in the following steps:
+Bootstrapping generates the initial call graph for TI. Once the call graph is generated, TI can perform Test Selection.
```
Think more info is needed here. Do you need to do this once or every time you set up TI for a different build stage? Also, use lower case on Test Selection.

aimurphy replied:
@hemanthmantri Could you help me clarify when the bootstrap is needed? Do you only need to this this the first time you enable TI on the pipeline, or every time you add it to any stage in a pipeline?

hemanthmantri replied:
Bootstrap is per repository. Every time a new repository is being built for the first time in a RunTests step, we need to follow the above process. From the user's perspective, nothing actually changes. They raise a PR, let the tests run and if they pass, they merge the PR. They don't have to know there's something called Bootstrap behind the scenes. The only reason we mention this is - only after the first PR has been merged, the subsequent PRs can take advantage of the Callgraph to select a subset of tests instead of all tests.

aimurphy replied:
@hemanthmantri Thank you! I revised it to: 
```
When you first enable Test Intelligence on a repo, you must run all tests to generate an initial call graph. This creates a baseline for test selection in future builds. To generate the initial call graph:

1. [Add a webhook trigger](../../../platform/11_Triggers/triggering-pipelines.md) to the pipeline that listens for PR merge events.
2. Open a PR against the pipeline's codebase. Make sure the build triggered by this PR runs all tests.
3. Merge the PR.
4. Wait while the pipeline executes. To monitor the build's progress, go to **Builds** and select the build that was started by the PR.
```

To clarify, the CI pipeline webhook trigger should listen for a PR to be *opened*, not *merged*? 
Should the steps be this instead?

```
1. [Add a webhook trigger](../../../platform/11_Triggers/triggering-pipelines.md) to the pipeline that listens for PRs to be opened against the pipeline's codebase.
2. Open a PR against the pipeline's codebase. Make sure the build triggered by this PR runs all tests.
3. Wait while the pipeline executes. To monitor the build's progress, go to **Builds** and select the build that was started by the PR.
4. If the tests pass and the build succeeds, merge the PR.
```

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
 
-* If this is the first CI stage in the Pipeline, in the Build stage settings, enable **Clone Codebase**.
-* If you have an existing Pipeline with a Build stage, click **Codebase** to configure your codebase settings**.**
+### Create the Build stage
```
I don't think this heading is accurate since you might already have a build stage where you want to add test intelligence. I think we can really simplify this topic by putting all the steps from Create the build stage through Configure the Run Tests step into a single procedure for setting up test intelligence.

aimurphy replied:
Will fix.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
-Displays the overview of Total Test, Tests Selected, Duration of the Selected Tests, and Time Saved.
+Displays an overview of **Total Tests**, number of **Selected Tests**, total **Duration** of all tests, and **Time Saved**.
+
+**Duration** reflects the sum of CPU time taken for all tests to complete. The values are collected as-is from the JUnit report, and they don't correspond with wall clock time. In contrast, the pipeline execution time is a measure of wall clock time. Therefore, it is possible that the **Duration** may exceed the total pipeline execution time.
```
"wall clock time"? Might not work for non-native English speakers.

aimurphy replied:
What would be a good term to use to distinguish between total CPU "run time" (As an aggregate of time taken for all tests, even if they were running simultaneously) and "real world" time (How long you would wait for the tests to finish)?

hemanthmantri replied:
`wall clock time` is a well-understood term in computer science and some test runners also print wall clock time in the output. Another term is `elapsed time` (https://www.ibm.com/docs/en/zos/2.2.0?topic=indicators-elapsed-time) but I have seen `wall clock time` or `wall time` used more often.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
 #### Test Selection Breakdown
 
-Displays the selection of Tests on the following metrics: changed code, changed tests, and new tests.
+Test Intelligence makes decisions about which tests to run by assessing changes to source files and test files in the codebase.
```
"makes decisions" - anthropomorphism

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
-Displays the selection of Tests on the following metrics: changed code, changed tests, and new tests.
+Test Intelligence makes decisions about which tests to run by assessing changes to source files and test files in the codebase.
+
+- **Correlated with Code Changes**: The number of tests that Test Intelligence ran based on changes in the codebase
+- **New Tests**: The number of tests that Test Intelligence ran because they are new
+- **Updated Tests**: The number of tests that Test Intelligence ran because the tests themselves were changed
```
Add periods at the end of each sentence.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+In subsequent pipeline runs, the test report shows stats corresponding to the tests selected by Test Intelligence for that run.
 
-Visualization shows why a specific test was selected and the reason behind every test selection. Click on any test (the purple node), and you can see all the classes and methods covered by this test. These are the changed classes and methods (denoted by the blue nodes), which led to the selection of that test.
+Select **Expand graph** to display the Test Intelligence Visualization. Visualization shows why a specific test was selected and the reason behind every test selection. Purple nodes represent tests. Select any test (purple node) to see all the classes and methods covered by that test. Blue nodes represent changed classes and methods that caused Test Intelligence to select that test.
```
```suggestion
Select **Expand graph** to display the Test Intelligence Visualization, which shows why a specific test was selected and the reason behind every test selection. Purple nodes represent tests. Select any test (purple node) to see all the classes and methods covered by that test. Blue nodes represent changed classes and methods that caused Test Intelligence to select that test.
```

aimurphy commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+Go to the pipeline where you want to enable Test Intelligence, and do one of the following:
 
-### Step 3: Define the Build Farm Infrastructure
+* If the pipeline doesn't already have a **Build** stage:
+   1. Select **Add Stage**, and then select **Build**.
+   2. Input a **Stage Name**, enable **Clone Codebase**, and then select **Set Up Stage**.
+* If the pipeline already has as **Build** stage, select **Codebase** to configure the pipeline's codebase settings.
```
I'm going to condense this into a requirements section with tabs. The procedure needs the build stage, but users enabling TI probably already have a pipeline with a build stage. Experienced users will care more about which languages are supported than how to add a build stage.

aimurphy commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -10,170 +10,187 @@ helpdocs_is_private: false
+   3. **Namespaces**: For .NET C# only, supply a comma-separated list of namespace prefixes that you want to test.
+1. Select **Apply Changes** to save the step settings, and then select **Save** to save the pipeline.
 
 ![](./static/set-up-test-intelligence-02.png)
```
It was probably there because this topic originally didn't describe how to complete each field, it just said to look at the screenshot.

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -1,180 +1,170 @@
+1. In the Pipeline Studio, select the **Build** stage, and then select the **Execution** tab.
+2. Select **Add Step**, select **Add Step** again, and then select **Run Tests** from the **Step Library**.
+3. Configure the [Run Tests step settings](../../ci-technical-reference/configure-run-tests-step-settings.md) as follows:
+   1. **Name**: Input a unique name for the step.
+   2. **Description**: Input a description for the step.
+   3. **Container Registry**: Select a container registry connector. This is the container registry, such as DockerHub, where Harness pulls the image on which it runs build commands. This is optional for stages that use Harness Cloud build infrastructure.
+   4. **Image**: The name of the Docker image to use for running the build commands. Enter a Fully Qualified Image Name (FQIN) when using a private container registry. This is optional for stages that use Harness Cloud build infrastructure.
+   5. **Language**: Select the source code language to build.
+   6. **Build Tool**: Select the build automation tool. Supported tools vary by **Language**. For example, Harness supports [Bazel](https://bazel.build/), [Maven](https://maven.apache.org/), and [Gradle](https://gradle.org/) for **Java** and [NET CLI](https://docs.microsoft.com/en-us/dotnet/core/tools/) and [Nunit](https://nunit.org/) for **.NET:** .
+   7. **Build Arguments**: Input the [arguments](../../ci-technical-reference/configure-run-tests-step-settings.md) for the build tool. These are used as input for the chosen build tool.
+   8. **Test Report Paths**: Input one or more test report paths.
```
Consider putting these items in an expandable section.

aimurphy replied:
I moved the details over to the Run Tests step settings reference topic where it makes more sense for that info to be (at least for now). If I decide to elimiate the reference topics later, I will probably change this into a series of more detailed top-level steps, rather than a single top-level step with a bunch of substeps.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -1,180 +1,170 @@
+   7. **Build Arguments**: Input the [arguments](../../ci-technical-reference/configure-run-tests-step-settings.md) for the build tool. These are used as input for the chosen build tool.
+   8. **Test Report Paths**: Input one or more test report paths.
+4. Expand **Additional Configuration** and complete these fields:
+   1. **Packages**: Leave blank or provide a comma-separated list of source code package prefixes, such as `com.company., io.company.migrations`. If blank, Harness auto-detects the packages.
+   2. **Test Annotations**: Provide a comma-separated list of test annotations used in unit testing. Any method with a specified annotation is treated as a test method. The defaults are: `org.junit.Test, org.junit.jupiter.api.Test, org.testing.annotations.Test`
+   3. **Namespaces**: For .NET C# only, supply a comma-separated list of namespace prefixes that you want to test.
```
Consider putting these items in an expandable section.

aimurphy replied:
This is an important part of the initial setup and I am concerned that putting it in an accordion makes it seem optional.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
```diff
@@ -1,180 +1,170 @@
-#### Visualization
-
-Visualization shows why a specific test was selected and the reason behind every test selection. Click on any test (the purple node), and you can see all the classes and methods covered by this test. These are the changed classes and methods (denoted by the blue nodes), which led to the selection of that test.
+* **Test Execution Overview:** Overview of **Total Tests**, number of **Selected Tests**, total **Duration** of all tests, and **Time Saved**.
+  * **Duration** reflects the sum of CPU time taken for all tests to complete. The values are collected as-is from the JUnit report, and they don't correspond with wall clock time. In contrast, the pipeline execution time is a measure of wall clock time. Therefore, it is possible that the **Duration** may exceed the total pipeline execution time.
+* **Test Execution Result:** Graphical representation of successful and failed tests.
+* **Test Selection Breakdown:** Test Intelligence analyzes changes to source files and test files in the codebase and then runs only those tests relevant to the detected changes. This section reports how many tests ran based on different the types of code changes included in this build.
+  * **Correlated with Code Changes**: The number of tests that ran due to changes in the codebase.
+  * **New Tests**: The number of tests that ran because they are new.
+  * **Updated Tests**: The number of tests that ran because there was a change to the actual test code or content.
+* **Test Execution:** Detailed list of all tests, including class methods and test methods.
+  * Initially, the list shows only failed tests. To see all tests, toggle **Show all Tests**.
+  * You can sort the list by failure rate, duration, and total tests. You can also expand test suites to see details about individual tests in that suite.
+* **Call Graph:** The call graph is initially generated after completing the bootstrap. In subsequent builds, the test report shows stats corresponding to the tests selected by Test Intelligence for that run.
+  * Select **Expand graph** to display the Test Intelligence Visualization, which shows why a specific test was selected and the reason behind every test selection.
+  * Purple nodes represent tests. Select any test (purple node) to see all the classes and methods covered by that test.
+  * Blue nodes represent changes to classes and methods that caused Test Intelligence to select that test.
```
Consider putting this info in a table or some other format. I looked at the rendered content in the preview build and the number of bullets is a little overwhelming and hard to parse.

aimurphy replied:
I will put them in accordions.

KimberlyFields commented on docs/continuous-integration/use-ci/set-up-test-intelligence/set-up-test-intelligence.md:
````diff
@@ -1,180 +1,170 @@
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+```
+```mdx-code-block
````
I looked at the rendered content in the preview build and I'm not about the use of tabs for these three requirements. I think of tabs as more useful in situations where there are many choices and the user only needs to pick the one they need. For example, for lists of parameters for different languages. The user many only use Go, so they know they just have to pick the Go tab for their instructions. Consider adding a sentence to guide the user to select each tab for details about each requirement or consider an alternate presentation.

aimurphy replied:
What about accordions?

I'm operating on the assumption that users will already have some kind of pipeline set up before they get to setting up test intelligence, so I don't want these prerequisite steps to overwhelm the topic's actual purpose.

KimberlyFields replied:
Accordions should work.

aimurphy commented:
- Moved each setting's details to a more appropriate location, in the Run Tests step settings reference topic. General clean up of Run Tests step settings topic.
- Moved the YAML code up into an accordion under the step settings. The YAML can only be used to configure the step, so they still need to generate the initial call graph. If they skip to the end (where the YAML originally was), they have a higher chance of missing the call graph step.
- Not sure what to do about the roman numerals for the call graph. I will undoubtedly be back in this topic in the future, so I will revisit it then.
- Use accordions for the requirements and test report sections.
- Noticed an unnecessarily repeated sentence - reduced it into one location in the intro.

aimurphy commented:
Rebased to resolve a merge conflict

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631

KimberlyFields commented on docs/continuous-integration/ci-technical-reference/configure-run-tests-step-settings.md:
```diff
@@ -1,157 +1,147 @@
-These settings specify the maximum resources used by the container at runtime.
+* **Always:** The kubelet queries the container image registry to resolve the name to an image digest every time the kubelet launches a container. If the kubelet encounters an exact digest cached locally, it uses its cached image; otherwise, the kubelet downloads (pulls) the image with the resolved digest, and uses that image to launch the container.
+* **If Not Present:** The image is pulled only if it isn't already present locally.
+* **Never:** The kubelet assumes that the image exists locally and doesn't try to pull the image.
```
A kubelet can't assume anything. Suggest revising the sentence to "The image is not pulled."

KimberlyFields commented on docs/continuous-integration/ci-technical-reference/configure-run-tests-step-settings.md:
```diff
@@ -1,157 +1,147 @@
+### Run as User
 
-Maximum memory that the container can use. You can express memory as a plain integer or as a fixed-point number using the suffixes `G` or `M`. You can also use the power-of-two equivalents `Gi` and `Mi`.
+Set the value to specify the user Id for all processes in the pod, running in containers. For more information, go to [Set the security context for a pod](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod).
```
```suggestion
Set the value to specify the user ID for all processes in the pod, running in containers. For more information, go to [Set the security context for a pod](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-pod).
```

KimberlyFields replied:
Not sure why you have a comma after "pod"

KimberlyFields reviewed (approved):
I left several new comments and suggestions, but I'm going to approve this PR with the understanding that you will address the comments.

KimberlyFields commented:
@aimurphy the preview build didn't show your most recent changes. Not sure why. I assume that you tested the updated format locally and that the new expandables and tabs work as intended.

bot-gitexp-user commented:
Preview environment: https://ci.pr.harness.io/pr-631
